## Optimización - Laboratorio 6

### Isabella Salgado 201730418
### Juan Pablo Naranjo 201730006

Se importan las librerías relevantes para el laboratorio

In [32]:
import numpy as np
import sympy
from sympy import *
import itertools
from itertools import combinations
import scipy as sp
from scipy.optimize import linprog
import warnings
warnings.filterwarnings('ignore')

# Punto 1

## a)

| Planta  |  |  Costo/Producto|   | 
|:-:|:-----------:|:-----:|:-----:|
|   |      1      |   2   |   3   |
| 1 |    \$30     | \$45  | \$38  | 
| 2 |    \$29     | \$41  | \$35  | 
| 3 |    \$32     | \$46  | \$40  | 
| 4 |    \$28     | \$42  | No    | 
| 5 |    \$29     | \$43  | No    | 

Sumatoria capacidades de las plantas = 3000 

Sumatoria producciones de los productos = 3000

## b)

El problema matemático a resolver es un problema de flujo con mínimo costo. Específciamente, es un problema de transporte. Las variables de decisión de este problema son las cantidades de productos nuevos que se van a producir. 

En este sentido, se usará la nomenclatura $x_{ij}$ para representar el producto $j$, producido en la planta $i$. Los productos $x_{ij}$ son las variables de decisión del problema de optimización.

Al ser un problema de flujo con mínimo costo, la función objetivo del problema de optimización es una función que relaciona los costos de transportar los productos $x_{ij}$ por cada arco de la red. Los costos asociados a cada arco están relacionados en la tabla del apartado a). Entonces, la función objetivo está dada por la siguiente expresión:

$$f(x) = 30x_{16} + 29x_{26} + 32x_{36} + 28x_{43} + 29x_{56} + 45x_{17} + 41x_{27} + 46x_{37} + 42x_{47} + 43x_{57} + 38x_{18} + 35x_{28} + 40x_{38}$$

Es importante mencionar que los productos 1, 2 y 3 se representaron en el problema con los números 6, 7 y 8, respectivamente, tal como se muestra en la representación gráfica adjunta a este cuaderno (véase el archivo titulado "grafoP1.png").

Las restricciones asociadas a este problema tienen indican que la diferencia entre la cantidad de flujo que sale de un nodo $i$ y la cantidad de flujo que entra al nodo $i$, tiene que ser igual a una cantidad $b_i$. Esta cantidad es un suministro si $b_i > 0$ o un requerimiento si $b_i < 0 $. En el caso de este problema, estos $b_i$ representan las capacidades de producción de cada planta (suministros) y la producción diaria de los productos (requerimientos). Adicionalmente, se tiene como restricción que el problema sea balanceado. Es decir, que la sumatoria de los requerimientos y de los suministros sean igual a cero. Esta condicicón se cumple para este problema, pues $\sum_{i=1}^{8}b_i = 0$.

Tomando en cuenta lo presentado anteriormente y la representación gráfica del problema, se plantean las siguientes restricciones:

Plantas \
$x_{16} + x_{17} + x_{18} = 400$ \
$x_{26} + x_{27} + x_{28} = 600$ \
$x_{36} + x_{37} + x_{38} = 400$ \
$x_{46} + x_{47} + x_{48} = 600$ \
$x_{56} + x_{57} + x_{58} = 1000$ 

Productos \
$- (x_{16} + x_{26} + x_{36} + x_{46} + x_{56}) = -1400$ \
$- (x_{17} + x_{27} + x_{37} + x_{47} + x_{57}) = -1000$ \
$- (x_{18} + x_{28} + x_{38} + x_{48} + x_{58}) = -600$ 

Problema balanceado \
$b_1 = 400$ \
$b_2 = 600$ \
$b_3 = 400$ \
$b_4 = 600$ \
$b_5 = 1000$ \
$b_6 = -1400$ \
$b_7 = -1000$ \
$b_8 = -600$

$\sum_{i=1}^{8}b_i = 0$

Finalmente, el planteamiento del problema de optimización es el siguiente: 

max $f(x) = 30x_{16} + 29x_{26} + 32x_{36} + 28x_{43} + 29x_{56} + 45x_{17} + 41x_{27} + 46x_{37} + 42x_{47} + 43x_{57} + 38x_{18} + 35x_{28} + 40x_{38}$ \
s.a.

$x_{16} + x_{17} + x_{18} = 400$ \
$x_{26} + x_{27} + x_{28} = 600$ \
$x_{36} + x_{37} + x_{38} = 400$ \
$x_{46} + x_{47} + x_{48} = 600$ \
$x_{56} + x_{57} + x_{58} = 1000$ \
$- (x_{16} + x_{26} + x_{36} + x_{46} + x_{56}) = -1400$ \
$- (x_{17} + x_{27} + x_{37} + x_{47} + x_{57}) = -1000$ \
$- (x_{18} + x_{28} + x_{38} + x_{48} + x_{58}) = -600$

## d)

In [30]:
f = np.array([30,29,32,28,29,45,41,46,42,43,38,35,40,0,0])
A = np.array([[1,0,0,0,0,1,0,0,0,0,1,0,0,0,0],
             [0,1,0,0,0,0,1,0,0,0,0,1,0,0,0],
             [0,0,1,0,0,0,0,1,0,0,0,0,1,0,0],
             [0,0,0,1,0,0,0,0,1,0,0,0,0,0,0],
             [0,0,0,0,1,0,0,0,0,1,0,0,0,0,0],
             [-1,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0],
             [0,0,0,0,0,-1,-1,-1,-1,-1,0,0,0,0,0],
             [0,0,0,0,0,0,0,0,0,0,-1,-1,-1,-1,-1]])
b = np.array([[400],[600],[400],[600],[1000],[-1400],[-1000],[-600]])

bounds = []

for i in range(15):
    bounds.insert(i,tuple((0,None)))

res = linprog(f, A_ub=None, b_ub=None, A_eq = A, b_eq = b, bounds = bounds, method='simplex', options={"disp": True})
print(res)

Optimization terminated successfully.
         Current function value: 105600.000000
         Iterations: 12
     con: array([0., 0., 0., 0., 0., 0., 0., 0.])
     fun: 105600.0
 message: 'Optimization terminated successfully.'
     nit: 12
   slack: array([], dtype=float64)
  status: 0
 success: True
       x: array([   0.,    0.,    0.,  400., 1000.,    0.,  400.,  400.,  200.,
          0.,  400.,  200.,    0.,    0.,    0.])


El costo de fabricación es de 105600. La fabricación por plantas queda de la siguiente forma:
- Planta 1: No fabrica productos
- Planta 2: 400 unidades del producto 1 y 1000 del producto 2
- Planta 3: 400 unidades del producto 1, 400 del producto 2 y 200 del producto 3
- Planta 4: 400 unidades del producto 2 y 200 del producto 3
- Planta 5: No fabrica productos

# Punto 2

## b)

A continuación se muestra el planeteamiento del problema para maximizar el número máximo de personas que se pueden evacuar del edificio en 1 minuto. A continuación se encuentran detalladas las capacidades máximas de personas de cada salón (nodo) del edificio:

1. Salón A: 20 personas 

2. Salón B: 25 personas

3. Salón C: 50 personas

4. Salón D: 30 personas

5. Salón E: 20 personas

6. Hall F: 80 personas

Adicionalmente, se creó un nodo fuente llamado s que inserta personas a los salones, y un nodo sumidero que sirve como el punto de encuentro de evacuación al que llegan todas las personas después de salir del edificio. Este nodo se llamó t. Para una mejor organización del problema, se le asignó un número a cada nodo, y se definió que el flujo que hay entre dos nodos $i$ y $j$ es $x_{ij}$. Estos flujos son las variables de decisión del problema de optimización.

El problema de máximo flujo pretende maximizar el flujo de la red. Esto quiere decir que se pretende o maximizar la cantidad de flujo que sale del nodo fuente o la cantidad de flujo que entra al nodo sumidero. Siguiendo la nomenclatura descrita anteriormente y que se puede apreciar en la imagen "grafoP2b.png", se plantea el problema como sigue:

Al lado de cada restricción se anota en paréntesis el número del nodo al que pertenece dicha restricción.

max $f(x) = x_{14} + x_{13} + x_{17} + x_{12} + x_{15} + x_{16} $ s.a.

$x_{14} + x_{13} + x_{17} + x_{12} + x_{15} + x_{16} - (x_{48} + x_{78} + x_{17} + x_{68}) = 0$ (Lo que sale del nodo fuente menos lo que entra al sumidero = 0) \
$x_{27} - x_{12} = 0$ (2)\
$x_{37} - x_{13} = 0$ (3)\
$x_{48} + x_{47} - (x_{14} + x_{74}) = 0$ (4)\
$x_{57} - x_{15} = 0$ (5)\
$x_{68} + x_{67} - (x_{16} + x_{76}) = 0$ (6) \
$(x_{78} + x_{74} + x_{76}) - (x_{47} + x_{67} + x_{37} + x_{17} + x_{27} + x_{57} + x_{67}) = 0$ (7)\
$x_{47} + x_{74} \leq 50$\
$x_{67} + x_{76} \leq 15$

$0 \leq x_{14} \leq 50$ \
$0 \leq x_{13} \leq 25$ \
$0 \leq x_{17} \leq 80$ \
$0 \leq x_{12} \leq 20$ \
$0 \leq x_{15} \leq 30$ \
$0 \leq x_{16} \leq 20$ \
$0 \leq x_{48} \leq 15$ \
$0 \leq x_{78} \leq 75$ \
$0 \leq x_{68} \leq 15$ \
$0 \leq x_{37} \leq 50$ \
$0 \leq x_{27} \leq 50$ \
$0 \leq x_{57} \leq 50$ \
$0 \leq x_{47} \leq 50$ \
$0 \leq x_{67} \leq 15$


In [59]:
f = np.array([[-1,-1,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0]])
A = np.array([[-1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0],
             [0,-1,0,0,0,0,0,1,0,0,0,0,0,0,0,0],
             [0,0,-1,0,0,0,0,0,1,1,0,0,0,-1,0,0],
             [0,0,0,-1,0,0,0,0,0,0,1,0,0,0,0,0],
             [0,0,0,0,-1,0,0,0,0,0,0,1,1,0,-1,0],
             [0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,0,1,1,1],
             [1,1,1,1,1,1,0,0,0,-1,0,0,-1,0,0,-1]])
b = np.zeros((7,1))

A_ub = np.array([[0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0],
                 [0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0]])

b_ub = np.array([[50], [15]])

bounds12 = (0,20)
bounds13 = (0,25)
bounds14 = (0,50)
bounds15 = (0,30)
bounds16 = (0,20)
bounds17 = (0,80)
bounds27 = (0,50)
bounds37 = (0,50)
bounds47 = (0,50)
bounds48 = (0,15)
bounds57 = (0,50)
bounds67 = (0,50)
bounds68 = (0,15)
bounds74 = (0,50)
bounds76 = (0,15)
bounds78 = (0,75)

bounds = [bounds12,bounds13,bounds14,bounds15,bounds16,bounds17,bounds27,bounds37,bounds47,bounds48,bounds57,bounds67,bounds68,bounds74,bounds76,bounds78]

res = linprog(f, A_ub=A_ub, b_ub=b_ub, A_eq = A, b_eq = b, bounds = bounds, method='simplex', options={"disp": True})
print(res)

Optimization terminated successfully.
         Current function value: -105.000000 
         Iterations: 26
     con: array([0., 0., 0., 0., 0., 0., 0.])
     fun: -105.0
 message: 'Optimization terminated successfully.'
     nit: 26
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([20. , 25. ,  0. , 30. ,  0. , 30. , 20. , 25. , 17.5, 15. , 30. ,
        0. , 15. , 32.5, 15. , 75. ])


No es posible evacuar todas las personas en 1 minuto.

## c)

A continuación se muestra el planeteamiento del problema para maximizar el número máximo de personas que se pueden evacuar del edificio en 2 minuto. Las capacidades máximas de personas de cada salón (nodo) del edificio son las mismas:

1. Salón A: 20 personas 

2. Salón B: 25 personas

3. Salón C: 50 personas

4. Salón D: 30 personas

5. Salón E: 20 personas

6. Hall F: 80 personas

El grafo para este problema se peude encontrar en la imagen "grafoP2c.png". El planteamiento se presenta a continuación:

Al lado de cada restricción se anota en paréntesis el número del nodo al que pertenece dicha restricción.

max $f(x) = x_{14} + x_{13} + x_{17} + x_{12} + x_{15} + x_{16} $ s.a.

$x_{14} + x_{13} + x_{17} + x_{12} + x_{15} + x_{16} - (x_{48} + x_{78} + x_{17} + x_{68}) = 0$ (Lo que sale del nodo fuente menos lo que entra al sumidero = 0) \
$x_{27} - x_{12} = 0$ (2)\
$x_{37} - x_{13} = 0$ (3)\
$x_{48} + x_{47} - (x_{14} + x_{74}) = 0$ (4)\
$x_{57} - x_{15} = 0$ (5)\
$x_{68} + x_{67} - (x_{16} + x_{76}) = 0$ (6) \
$(x_{78} + x_{74} + x_{76}) - (x_{47} + x_{67} + x_{37} + x_{17} + x_{27} + x_{57} + x_{67}) = 0$ (7)\
$x_{47} + x_{74} \leq 100$ \
$x_{67} + x_{76} \leq 30$

$0 \leq x_{14} \leq 50$ \
$0 \leq x_{13} \leq 25$ \
$0 \leq x_{17} \leq 80$ \
$0 \leq x_{12} \leq 20$ \
$0 \leq x_{15} \leq 30$ \
$0 \leq x_{16} \leq 20$ \
$0 \leq x_{48} \leq 30$ \
$0 \leq x_{78} \leq 150$ \
$0 \leq x_{68} \leq 30$ \
$0 \leq x_{37} \leq 100$ \
$0 \leq x_{27} \leq 100$ \
$0 \leq x_{57} \leq 100$ \
$0 \leq x_{47} \leq 100$ \
$0 \leq x_{67} \leq 30$

In [58]:
f = np.array([[-1,-1,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0]])
A = np.array([[-1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0],
             [0,-1,0,0,0,0,0,1,0,0,0,0,0,0,0,0],
             [0,0,-1,0,0,0,0,0,1,1,0,0,0,-1,0,0],
             [0,0,0,-1,0,0,0,0,0,0,1,0,0,0,0,0],
             [0,0,0,0,-1,0,0,0,0,0,0,1,1,0,-1,0],
             [0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,0,1,1,1],
             [1,1,1,1,1,1,0,0,0,-1,0,0,-1,0,0,-1]])
b = np.zeros((7,1))

A_ub = np.array([[0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0],
                 [0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0]])

b_ub = np.array([[100], [30]])

bounds12 = (0,20)
bounds13 = (0,25)
bounds14 = (0,50)
bounds15 = (0,30)
bounds16 = (0,20)
bounds17 = (0,80)
bounds27 = (0,100)
bounds37 = (0,100)
bounds47 = (0,100)
bounds48 = (0,30)
bounds57 = (0,100)
bounds67 = (0,100)
bounds68 = (0,30)
bounds74 = (0,100)
bounds76 = (0,30)
bounds78 = (0,150)

bounds = [bounds12,bounds13,bounds14,bounds15,bounds16,bounds17,bounds27,bounds37,bounds47,bounds48,bounds57,bounds67,bounds68,bounds74,bounds76,bounds78]

res = linprog(f, A_ub=A_ub, b_ub=b_ub, A_eq = A, b_eq = b, bounds = bounds, method='simplex', options={"disp": True})
print(res)

Optimization terminated successfully.
         Current function value: -210.000000 
         Iterations: 24
     con: array([0., 0., 0., 0., 0., 0., 0.])
     fun: -210.0
 message: 'Optimization terminated successfully.'
     nit: 24
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([ 20. ,  25. ,  35. ,  30. ,  20. ,  80. ,  20. ,  25. ,  52.5,
        30. ,  30. ,  10. ,  30. ,  47.5,  20. , 150. ])


No es posible evacuar la planta en 2 minutos, ya que salen 205 personas.

# Punto 3

## a)

De los puntos anteriores, se observa que hay una relación lineal entre los minutos y la cantidad de personas que evacúan la planta. Cada minuto salen 105 personas de la planta. Con esta relación se puede establecer cuantos minutos tomaría evacuar la planta completa $x$, es decir, que salgan 225 personas.

$x = 225 personas * \frac{1 minuto(s)}{105 personas} \longrightarrow x = 2.143 minutos$ 

Evacuar toda la planta toma 2.143 minutos.

## b)

Se soluciona el problema de programación lineal aumentando en 10 las puertas C y F.

In [69]:
f = np.array([[-1,-1,-1,-1,-1,-1,0,0,0,0,0,0,0,0,0,0]])
A = np.array([[-1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0],
             [0,-1,0,0,0,0,0,1,0,0,0,0,0,0,0,0],
             [0,0,-1,0,0,0,0,0,1,1,0,0,0,-1,0,0],
             [0,0,0,-1,0,0,0,0,0,0,1,0,0,0,0,0],
             [0,0,0,0,-1,0,0,0,0,0,0,1,1,0,-1,0],
             [0,0,0,0,0,-1,-1,-1,-1,0,-1,-1,0,1,1,1],
             [1,1,1,1,1,1,0,0,0,-1,0,0,-1,0,0,-1]])
b = np.zeros((7,1))

A_ub = np.array([[0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0],
                 [0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0]])

b_ub = np.array([[50], [15]])

bounds12 = (0,20)
bounds13 = (0,25)
bounds14 = (0,50)
bounds15 = (0,30)
bounds16 = (0,20)
bounds17 = (0,80)
bounds27 = (0,50)
bounds37 = (0,50)
bounds47 = (0,50)
bounds57 = (0,50)
bounds67 = (0,50)
bounds74 = (0,50)
bounds76 = (0,15)

bounds48 = (0,25)
bounds68 = (0,15)
bounds78 = (0,85)

bounds1 = [bounds12,bounds13,bounds14,bounds15,bounds16,bounds17,bounds27,bounds37,bounds47,bounds48,bounds57,bounds67,bounds68,bounds74,bounds76,bounds78]

res1 = linprog(f, A_ub=A_ub, b_ub=b_ub, A_eq = A, b_eq = b, bounds = bounds1, method='simplex', options={"disp": True})
print(res1)

Optimization terminated successfully.
         Current function value: -125.000000 
         Iterations: 26
     con: array([0., 0., 0., 0., 0., 0., 0.])
     fun: -125.0
 message: 'Optimization terminated successfully.'
     nit: 26
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([20. , 25. ,  0. , 30. , 10. , 40. , 20. , 25. , 12.5, 25. , 30. ,
        0. , 25. , 37.5, 15. , 75. ])


In [54]:
bounds48 = (0,25)
bounds68 = (0,15)
bounds78 = (0,85)

bounds1 = [bounds12,bounds13,bounds14,bounds15,bounds16,bounds17,bounds27,bounds37,bounds47,bounds48,bounds57,bounds67,bounds68,bounds74,bounds76,bounds78]

res1 = linprog(f, A_ub=A_ub, b_ub=b_ub, A_eq = A, b_eq = b, bounds = bounds1, method='simplex', options={"disp": True})
print(res1)

Optimization terminated successfully.
         Current function value: -125.000000 
         Iterations: 26
     con: array([0., 0., 0., 0., 0., 0., 0.])
     fun: -125.0
 message: 'Optimization terminated successfully.'
     nit: 26
   slack: array([0., 0.])
  status: 0
 success: True
       x: array([20. , 25. ,  0. , 30. ,  0. , 50. , 20. , 25. , 12.5, 25. , 30. ,
        0. , 15. , 37.5, 15. , 85. ])


Se decide expandir las puertas C y F ya que son las que dan mayor número de personas evacuadas en un minuto, con 125 personas.

## c)

Al haber la nueva relación lineal entre los minutos y la cantidad de personas que evacúan la planta: cada minuto salen 125 personas de la planta, se puede establecer cuantos minutos tomaría evacuar la planta completa $x$, es decir, que salgan 225 personas.

$x = 225 personas * \frac{1 minuto(s)}{125 personas} \longrightarrow x = 1.8 minutos$ 

Evacuar toda la planta toma 1.8 minutos al expandir 2 puertas.